src/box/alter.cc

```diff
@@ -2158,6 +2168,22 @@ on_replace_dd_space(struct trigger * /* trigger */, void *event)
 			ER_CREATE_SPACE : ER_ALTER_SPACE;
 		def = space_def_new_from_tuple(new_tuple, errcode, region);
 	}
+	if ((old_space != NULL && space_is_meta_temporary(old_space)) ||
+	    (def != NULL && space_opts_is_meta_temporary(&def->opts))) {
+		txn_stmt_mark_as_temporary(txn, stmt);
+	} else {
+		/*
+		 * box_process1() bypasses the read-only check for the
+		 * _space system space because there it's not yet known
+		 * if the related space is meta-temporary. Perform the
+		 * check here if the space isn't meta-temporary and
+		 * the statement was issued by this replica.
+		 */
+		if (stmt->row->replica_id == 0 &&
+		    recovery_state != INITIAL_RECOVERY &&
+		    box_check_writable() != 0)
+			return -1;
```

Comment:
This is untrivial enough to be in a separate function. Duplicating this big comment doesn't look good if anything changes.

Reply:
No, it will overcomplicate the related code, considering that
1. this new function will only be usable in 2 places
2. will need an extra `if (...) return -1;`
3. coming up with a good name for such a function will be impossible and anybody reading the code will have to go and look at it's source anyway

Reply:
> this new function will only be usable in 2 places

It is not about number of places. With this kind of code blocks 2 is already too many. It is more about this place being too hard to explain even once, leave alone twice. If anything changes here, the chances are high the second place will go unnoticed.

> will need an extra if (...) return -1;

Oh please, this is DDL and you know that 1 `if` here won't matter at all.

> coming up with a good name for such a function will be impossible

*"Impossible"* is quite a strong statement. Did you try?
```Diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index a18bb65898..4f4dd9b3e1 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -2088,6 +2088,25 @@ space_check_alter(struct space *old_space, struct space_def *new_space_def)
 	return 0;
 }
 
+/*
+ * box_process1() bypasses the read-only check for the _space system space
+ * because there it's not yet known if the related space is temporary. Perform
+ * the check here if the space isn't temporary and the statement was issued by
+ * this replica.
+ */
+static int
+filter_temporary_ddl_stmt(struct txn *txn, const struct space_def *def)
+{
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	if (space_opts_is_temporary(&def->opts)) {
+		txn_stmt_mark_as_temporary(txn, stmt);
+		return 0;
+	}
+	if (stmt->row->replica_id == 0 && recovery_state != INITIAL_RECOVERY)
+		return box_check_writable();
+	return 0;
+}
+
 /**
  * A trigger which is invoked on replace in a data dictionary
  * space _space.
@@ -2171,22 +2190,9 @@ on_replace_dd_space(struct trigger * /* trigger */, void *event)
 			ER_CREATE_SPACE : ER_ALTER_SPACE;
 		def = space_def_new_from_tuple(new_tuple, errcode, region);
 	}
-	if ((old_space != NULL && space_is_temporary(old_space)) ||
-	    (def != NULL && space_opts_is_temporary(&def->opts))) {
-		txn_stmt_mark_as_temporary(txn, stmt);
-	} else {
-		/*
-		 * box_process1() bypasses the read-only check for the
-		 * _space system space because there it's not yet known
-		 * if the related space is temporary. Perform the
-		 * check here if the space isn't temporary and
-		 * the statement was issued by this replica.
-		 */
-		if (stmt->row->replica_id == 0 &&
-		    recovery_state != INITIAL_RECOVERY &&
-		    box_check_writable() != 0)
-			return -1;
-	}
+	if (filter_temporary_ddl_stmt(txn, old_space != NULL ?
+				      old_space->def : def) != 0)
+		return -1;
 	if (new_tuple != NULL && old_space == NULL) { /* INSERT */
 		if (def == NULL)
 			return -1;
@@ -2516,21 +2522,8 @@ on_replace_dd_index(struct trigger * /* trigger */, void *event)
 	struct space *old_space = space_cache_find(id);
 	if (old_space == NULL)
 		return -1;
-	if (space_is_temporary(old_space)) {
-		txn_stmt_mark_as_temporary(txn, stmt);
-	} else {
-		/*
-		 * box_process1() bypasses the read-only check for the
-		 * _index system space because there it's not yet known
-		 * if the related space is temporary. Perform the
-		 * check here if the space isn't temporary and
-		 * the statement was issued by this replica.
-		 */
-		if (stmt->row->replica_id == 0 &&
-		    recovery_state != INITIAL_RECOVERY &&
-		    box_check_writable() != 0)
-			return -1;
-	}
+	if (filter_temporary_ddl_stmt(txn, old_space->def) != 0)
+		return -1;
 	if (old_space->def->opts.is_view) {
 		diag_set(ClientError, ER_ALTER_SPACE, space_name(old_space),
 			  "can not add index on a view");
```

Reply:
done